### PR TITLE
[site] Sidebar ToC for documentation books

### PR DESCRIPTION
**Pull request by hcallahan-lowrisc**
~~DRAFT PR, UNFINISHED~~
READY FOR REVIEW

I have been sorely-missing navigation options within a page on the docs site, so I spent a bit of time working on this. 
Feedback and comments on the design, style & function very welcome.
At this point, I think the design/styling is just okay. Don't love it or hate it. If anyone wants to restyle, it should be easy to do from here.

[LIVE_DEMO - staging.opentitan.org](https://staging.opentitan.org/book/hw/ip/otbn/index.html)

Features
- Based on css/js from https://github.com/JorelAli/mdBook-pagetoc
- Sidebar ToC, clickable nav links, auto-highlight the closest heading as you scroll within a page
- ToC auto-scrolls so the current heading in the window is always visible.
- Indentation lines to show subheading levels, with extra styling if you hover.
- Active target (most recently clicked link) highlighted with a horizontal (orange-ish) bar

Rough Edges (that I know of)
- ~~Sometimes the last heading on a page is left-off the ToC~~ FIXED, not updated in staging. 
- Some pages have waaay too many headings (e.g. [this](https://staging.opentitan.org/book/hw/ip/otbn/doc/isa.html)), so we need some way to filter the list to a manageable length. Some filtering is done currently (eg. for register maps, the subsections are removed).
---

Screenshot from 2023-05-25 20-39-14

### CODE REVIEW

**Comment by andreaskurth** [comment]
This will be very useful, thanks for working on it! :+1:

**Comment by pamaury** [comment]
I cannot really comment on the code but I am also really missing this so I really hope we can merge it.

**Comment by luismarques** [comment]
This is looking nice! I'm also not the best person to review web stuff.

**Comment by marnovandermaas** [comment]
Thanks for working on this. A couple of things I noticed, the highlight of a heading goes through behind the table of content:
image
Also, some pages don't have a table of contents:
image

Great start on this, and I definitely want this in there!

**Comment by jwnrt** [comment]
Thanks for working on this @hcallahan-lowrisc, I think a sidebar will be useful

Here's a couple of issues I've seen in the live demo:

* Some headings are missing from the TOC, e.g. "Check CI" and "Debugging a failed verilator test" in this screenshot:
  <details>
  <summary>Show/hide</summary>

  image

  </details>
* The main content is shifted all the way to the right, probably related to the flex problem you've mentioned already (firefox):
  <details>
  <summary>Show/hide></summary>
     
  image
  
  </details>

**Comment by hcallahan-lowrisc** [comment]
> Thanks for working on this. A couple of things I noticed, the highlight of a heading goes through behind the table of content: image 
> 

Yeah I chose to do this, I thought it looked quite clean. WDYT?

**Comment by marnovandermaas** [comment]
> > Thanks for working on this. A couple of things I noticed, the highlight of a heading goes through behind the table of content: image
> 
> @marnovandermaas Yeah I chose to do this, I thought it looked quite clean. WDYT?

It looked a bit weird to me at first, but I'm warming up to it now :wink:

**Comment by moidx** [comment]
Thanks @hcallahan-lowrisc! Looks great! In order to make the TOC work in mobile layout, would it make sense to think about moving it to the side panel below the current links, or what else would make sense to do?

**Comment by hcallahan-lowrisc** [comment]
Still need to fix the missing headings issue, but the previous commit added some logic for displaying a scrollbar when the list is longer than the page, and fixed the centering issue. Still WIP.

**Comment by hcallahan-lowrisc** [comment]
> Thanks @hcallahan-lowrisc! Looks great! In order to make the TOC work in mobile layout, would it make sense to think about moving it to the side panel below the current links, or what else would make sense to do?

@moidx Thanks! From my experience, most websites handle the mobile experience by just not showing a ToC, and keeping the list of pages under a hamburger menu or similar (eg. https://rtfm.newae.com/Targets/CW310%20Bergen%20Board/). I think the UX might be a bit confusing with one megalist.
I think it's a good suggestion though, and it's definitely important to keep the mobile experience in mind. How about I create a follow-up issue to consider adding more navigation options on mobile?

**Comment by hcallahan-lowrisc** [comment]
@HU90m @jwnrt @marnovandermaas Would any of you mind taking a look at this? There is 200-odd lines of javascript, but well-commented so it shouldn't be too much of a drag.
The staging site currently has a build of this PR on it (not quite the latest, there are some very minor differences to head of this PR).

File: site/book-theme/pagetoc.js [hunk 1]
```diff
@@ -43,6 +43,31 @@ var updateDynamicHighlight = function() {
...
+        newurl = window.location.href;
+    } else {
+        // "hashchange" event
+        newurl = event.newURL;
+    }
+    Array.prototype.forEach.call(document.getElementsByClassName("header"), function(el) {
+        if (new URL(el.href).hash == new URL(newurl).hash) {
+            document.documentElement.style.setProperty(
+                '--target-height',
+                el.getBoundingClientRect().height + 20 + "px"
```
Comment (jwnrt) [inline site/book-theme/pagetoc.js:69]: Where does this 20 come from?

File: site/book-theme/pagetoc.js [hunk 2]
```diff
@@ -68,6 +68,22 @@ var set_target_highlight = function(event) {
...
 window.addEventListener("load", set_target_highlight);
 
+/* Set the "height" style of pagetoc conditionally.
+ * - auto    -> for short lists that don't overflow, limit the height of the pagetoc, disables scrollbar.
+ * - limited -> content overflows the element, and therefore scrolling is enabled */
+var set_pagetoc_height = function() {
+    let el = document.getElementsByClassName("pagetoc")[0];
+    el.style.height = "auto";
+    let pagetoc_height = el.getBoundingClientRect().height;
+    let window_height = window.innerHeight - document.documentElement.style.getPropertyValue('--menu-bar-height') - 200;
```
Comment (jwnrt) [inline site/book-theme/pagetoc.js:84]: 200?

**Review by jwnrt — COMMENTED** [review]
Thanks again for this @hcallahan-lowrisc. Had a bit of time for a quick review:

* Code looks good, but I only reviewed your changes on top of pagetoc and only in shallow depth.
* Design-wise, I like the coloured border but personally don't think it's right for the docs.
* I think there should be more space between the content and the TOC box, it feels quite close right now.
* Do you think we need to clear the WTFPL code with anyone?

**Comment by hcallahan-lowrisc** [comment]
> * Design-wise, I like the coloured border but personally don't think it's right for the docs.

I agree, but at this point I'm not really sure how best to restyle it. I think its okay for now, so how about we come back to this as a follow-up task?

> * I think there should be more space between the content and the TOC box, it feels quite close right now.

If we change the border, this would need to change too. Maybe roll it into the follow-up? If you want to play with it locally, you would [need to change this](https://github.com/lowRISC/opentitan/pull/18411/files#diff-110221d9954dcad4db44ba94adf8b348ca4102dc8db95bde7b61f5422cd96cfcR17).

> * Do you think we need to clear the WTFPL code with anyone?

I included the files but relicensed to the standard OT Apache-2.0, so it shouldn't be a problem. I also added a comment giving credit to the original author/source.

File: site/book-theme/pagetoc.js [hunk 3]
```diff
@@ -15,26 +15,33 @@ Array.prototype.forEach.call(document.getElementsByClassName("pagetoc")[0].child
...
+/* The following functionality highlights the pagetoc entry of the highest visible heading on the page.
+ * This gives the pagetoc the dynamic highlighting behaviour as you scroll the page. */
+var updateDynamicHighlight = function() {
     var id;
-    var elements = document.getElementsByClassName("header");
+    let elements = document.getElementsByClassName("header");
+    // Set id == the highest "header" element visible in the window.
     Array.prototype.forEach.call(elements, function(el) {
-        if (window.pageYOffset >= el.offsetTop) {
+        if ((window.pageYOffset + 150) >= el.offsetTop) {
```
[the remark this answers is not in the document]
Comment (hcallahan-lowrisc) [reply site/book-theme/pagetoc.js:25]: See commit message for https://github.com/lowRISC/opentitan/pull/18411/commits/8fab966efce956b0b43ee5660a1a87e1f4615024

File: site/book-theme/pagetoc.js (hunk 1 above)
Comment (hcallahan-lowrisc) [reply site/book-theme/pagetoc.js:69]: In combination with [margin-top](https://github.com/lowRISC/opentitan/pull/18411/files#diff-270e09b784846a990b49d7c5553cbc3097ab1272ebd9e95804db697dd0e65e9fR78), this makes the highlight bar larger than the size of the heading itself, with approximately-equal top and bottom margins.

File: site/book-theme/pagetoc.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (hcallahan-lowrisc) [reply site/book-theme/pagetoc.js]: Done

**Review by marnovandermaas — APPROVED** [review]
Thanks for these changes Harry. I reviewed the preview on staging site and it looks good to me. Styling changes can be added in the future through follow up PRs. Can you make sure to create an issue to track the long table of contents?

**Pull request merged**